runtime/testy/runtime.py

```diff
@@ -103,17 +117,111 @@ def startStateManager(badThingsQueue, stateQueue, runtimePipe):
   except Exception:
     badThingsQueue.put(BadThing(sys.exc_info(), None))
 
-def processFactory(badThingsQueue, stateQueue):
-  def spawnProcessHelper(processName, helper):
+def processFactory(badThingsQueue, stateQueue, stdoutRedirect = None):
+  def spawnProcessHelper(processName, helper, *args):
     pipeToChild, pipeFromChild = multiprocessing.Pipe()
     if processName != PROCESS_NAMES.STATE_MANAGER:
       stateQueue.put([SM_COMMANDS.ADD, [processName, pipeToChild]], block=True)
       pipeFromChild.recv()
-    newProcess = multiprocessing.Process(target=helper, name=processName.value, args=(badThingsQueue, stateQueue, pipeFromChild))
+    newProcess = multiprocessing.Process(target=helper, name=processName.value, args=[badThingsQueue, stateQueue, pipeFromChild] + list(args))
     allProcesses[processName] = newProcess
     newProcess.daemon = True
     newProcess.start()
   return spawnProcessHelper
 
+def testTime():
+  # Normally dangerous. Allowed here because we put testing code there.
+  import studentCode
+
+  testNameRegex = re.compile(".*_setup")
+  testNames = [testName[:-len("_setup")] for testName in dir(studentCode) if testNameRegex.match(testName)]
+
+  failCount = 0
+  failedTests = []
+
+  for testName in testNames:
+    testFileName = "%s_output" % (testName,)
+    with open(testFileName, "w") as testOutput:
+      # Wrapper to flush after every write. This ensures file is written to in the same
+      # order that the 'print' function is called.
+      # https://mail.python.org/pipermail/tutor/2003-November/026645.html
+      class Unbuffered:
+        def __init__(self, stream):
+          self.stream = stream
+        def write(self, data):
+          self.stream.write(data)
+          self.stream.flush()
+        def __getattr__(self, attr):
+          return getattr(self.stream, attr)
+      # Will overwrite stdout for all spawned processes
+      # To print to terminal, comment out this line.
+      sys.stdout = Unbuffered(testOutput)
+
+      allProcesses.clear()
+
+      badThingsQueue = multiprocessing.Queue()
+      stateQueue = multiprocessing.Queue()
+      spawnProcess = processFactory(badThingsQueue, stateQueue, sys.stdout)
+      restartCount = 0
+
+      try:
+        spawnProcess(PROCESS_NAMES.STATE_MANAGER, startStateManager)
+        while True:
+          if restartCount >= 3:
+            break
+          spawnProcess(PROCESS_NAMES.STUDENT_CODE, runStudentCode, testName, 3)
```

Comment:
I dont feel like there is anything preventing us from merging this. 

The old runtime also only spawned two processes. This new function is a strict upgrade of the previous: it allows us to test more, more thoroughly, and more independently. 

Adding socket integration needs to be part of the socket integration commit. 

The command line arguments to toggle are in a commit Christene Zheng is working on. The functionality can still be readily simulated by commenting out one line of code. Furthermore, i dont see a way the old runtime code will be useful, other than running infinitely (which we also dont want if we want a test to finish.)